Pinot Server fails to start in offline/air-gapped environments due to eager evaluation of NetUtils.getHostAddress()

### Current Behavior

When attempting to start the Pinot Server (_version 1.4.0_) in an environment with no outbound internet access (e.g., a firewalled Docker Swarm overlay network), the server fails to initialize.

The startup process throws a java.io.UncheckedIOException: java.net.SocketException: Network is unreachable. This occurs because NetUtils.getHostAddress() is called, which attempts to open a socket to an external address (which is now 8.8.8.8) to determine the host's IP, failing in an offline environment.

### Steps to Reproduce

Configure a Pinot Server (v1.4.0) in an environment that has no access to the public internet (e.g., a Docker container with networking restricted, or a fully air-gapped VM).

Attempt to start the server using the standard StartServer command. **Note that the failure occurs even when explicitly providing the -serverHost parameter** (which is expected to set `pinot.server.netty.host` and bypass the network lookup)

The server process will fail during initialization.

Observe the stack trace (see below).

### Error Log / Stack Trace

The following exception is thrown during startup:

```
2025/11/05 10:35:20.073 INFO [StartServerCommand] [main] Executing command: StartServer -clusterName [my-cluster] -serverHost pinot-server-node1 -serverPort 8098 -serverAdminPort 8097 -serverGrpcPort 8090 -serverMultistageServerPort 0 -serverMultistageRunnerPort 0 -dataDir /var/lib/pinot/server -segmentDir /var/lib/pinot/segments -zkAddress zookeeper1:2181
2025/11/05 10:35:20.078 INFO [StartServiceManagerCommand] [main] Executing command: StartServiceManager -clusterName [my-cluster] -zkAddress zookeeper1:2181 -port -1 -bootstrapServices []
2025/11/05 10:35:20.079 INFO [StartServiceManagerCommand] [main] Starting a Pinot [SERVICE_MANAGER] at 0.217s since launch
2025/11/05 10:35:20.081 INFO [StartServiceManagerCommand] [main] Started Pinot [SERVICE_MANAGER] instance [ServiceManager_pinot-server-node1_-1] at 0.219s since launch
2025/11/05 10:35:20.082 INFO [StartServiceManagerCommand] [Start a Pinot [SERVER]] Starting a Pinot [SERVER] at 0.22s since launch
2025/11/05 10:35:20.354 ERROR [StartServiceManagerCommand] [Start a Pinot [SERVER]] Failed to start a Pinot [SERVER] at 0.493 since launch
java.io.UncheckedIOException: java.net.SocketException: Network is unreachable
	at java.base/sun.nio.ch.DatagramSocketAdaptor.connect(DatagramSocketAdaptor.java:120)
	at java.base/java.net.DatagramSocket.connect(DatagramSocket.java:474)
	at org.apache.pinot.spi.utils.NetUtils.getHostAddress(NetUtils.java:62)
	at org.apache.pinot.server.starter.helix.BaseServerStarter.init(BaseServerStarter.java:198)
	at org.apache.pinot.tools.service.PinotServiceManager.startServer(PinotServiceManager.java:166)
	at org.apache.pinot.tools.service.PinotServiceManager.startRole(PinotServiceManager.java:97)
	at org.apache.pinot.tools.admin.command.StartServiceManagerCommand$1.lambda$run$0(StartServiceManagerCommand.java:267)
	at org.apache.pinot.tools.admin.command.StartServiceManagerCommand.startPinotService(StartServiceManagerCommand.java:293)
	at org.apache.pinot.tools.admin.command.StartServiceManagerCommand$1.run(StartServiceManagerCommand.java:267)
Caused by: java.net.SocketException: Network is unreachable
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:579)
	at java.base/sun.nio.ch.DatagramChannelImpl.connect(DatagramChannelImpl.java:1249)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.connectInternal(DatagramSocketAdaptor.java:91)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.connect(DatagramSocketAdaptor.java:118)
	... 8 more
```

### Analysis & Root Cause

The root cause is in org.apache.pinot.server.starter.helix.BaseServerStarter.init (line 198 in version 1.4.0, or [line 205 on master](https://github.com/apache/pinot/blob/master/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java#L205)):

```java
_hostname = _serverConf.getProperty(Helix.KEY_OF_SERVER_NETTY_HOST,
        _serverConf.getProperty(Helix.SET_INSTANCE_ID_TO_HOSTNAME_KEY, false) ? NetUtils.getHostnameOrAddress()
            : NetUtils.getHostAddress());
```

The issue is that the second argument (the default value) to _serverConf.getProperty() is evaluated eagerly.

This means that NetUtils.getHostAddress() (or NetUtils.getHostnameOrAddress()) is always called, even if the user has explicitly provided a value for `pinot.server.netty.host` (Helix.KEY_OF_SERVER_NETTY_HOST) in their configuration to avoid this network lookup.

In an offline environment, this eager call to NetUtils.getHostAddress() fails with Network is unreachable, preventing the server from starting.

### Proposed Solution

I suggest a two-part solution:

1. Primary Fix (Lazy Evaluation): Refactor the logic in BaseServerStarter.init to only call NetUtils if Helix.KEY_OF_SERVER_NETTY_HOST is not already defined in the configuration.

2. Robustness Fix (NetUtils): It would also be beneficial to make NetUtils.getHostAddress() more robust. Instead of throwing an exception if the default probe address (e.g., 8.8.8.8) is unreachable, it could log a warning and fall back to the first available non-loopback IP address.

3. (Addition): The probe address(es) should be customizable.

### Environment

- *Pinot Version*: 1.4.0
- *Java Version*: Amazon Corretto 17 
- *Deployment*: Docker Swarm (on an internal overlay network)
- *OS*: Ubuntu 24.04

--- a/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java
+++ b/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java
@@ -219,9 +219,7 @@ public abstract class BaseServerStarter implements ServiceStartable {
 
     setupHelixSystemProperties();
     _listenerConfigs = ListenerConfigUtil.buildServerAdminConfigs(_serverConf);
-    _hostname = _serverConf.getProperty(Helix.KEY_OF_SERVER_NETTY_HOST,
-        _serverConf.getProperty(Helix.SET_INSTANCE_ID_TO_HOSTNAME_KEY, false) ? NetUtils.getHostnameOrAddress()
-            : NetUtils.getHostAddress());
+    _hostname = getServerHostname(_serverConf);
     // Override multi-stage query runner hostname if not set explicitly
     if (!_serverConf.containsKey(CommonConstants.MultiStageQueryRunner.KEY_OF_QUERY_RUNNER_HOSTNAME)) {
       _serverConf.setProperty(CommonConstants.MultiStageQueryRunner.KEY_OF_QUERY_RUNNER_HOSTNAME, _hostname);
@@ -321,6 +319,17 @@ public abstract class BaseServerStarter implements ServiceStartable {
    */
   protected QueryOperatorFactoryProvider createQueryOperatorFactoryProvider(PinotConfiguration serverConf) {
     return DefaultQueryOperatorFactoryProvider.INSTANCE;
+  }
+
+  @VisibleForTesting
+  static String getServerHostname(PinotConfiguration serverConf)
+      throws IOException {
+    String configuredHost = serverConf.getProperty(Helix.KEY_OF_SERVER_NETTY_HOST);
+    if (configuredHost != null) {
+      return configuredHost;
+    }
+    return serverConf.getProperty(Helix.SET_INSTANCE_ID_TO_HOSTNAME_KEY, false) ? NetUtils.getHostnameOrAddress()
+        : NetUtils.getHostAddress();
   }
 
   /**

--- a/pinot-server/src/test/java/org/apache/pinot/server/starter/helix/BaseServerStarterTest.java
+++ b/pinot-server/src/test/java/org/apache/pinot/server/starter/helix/BaseServerStarterTest.java
@@ -1,0 +1,79 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.server.starter.helix;
+
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants.Helix;
+import org.apache.pinot.spi.utils.NetUtils;
+import org.mockito.MockedStatic;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.never;
+import static org.testng.Assert.assertEquals;
+
+public class BaseServerStarterTest {
+  @Test
+  public void testConfiguredServerHostDoesNotResolveDefaultHost()
+      throws Exception {
+    PinotConfiguration serverConf = new PinotConfiguration();
+    serverConf.setProperty(Helix.KEY_OF_SERVER_NETTY_HOST, "pinot-server-node1");
+
+    try (MockedStatic<NetUtils> netUtils = mockStatic(NetUtils.class)) {
+      netUtils.when(NetUtils::getHostAddress).thenThrow(new AssertionError("Should not resolve host address"));
+      netUtils.when(NetUtils::getHostnameOrAddress).thenThrow(new AssertionError("Should not resolve hostname"));
+
+      assertEquals(BaseServerStarter.getServerHostname(serverConf), "pinot-server-node1");
+
+      netUtils.verify(NetUtils::getHostAddress, never());
+      netUtils.verify(NetUtils::getHostnameOrAddress, never());
+    }
+  }
+
+  @Test
+  public void testServerHostFallsBackToHostAddress()
+      throws Exception {
+    PinotConfiguration serverConf = new PinotConfiguration();
+
+    try (MockedStatic<NetUtils> netUtils = mockStatic(NetUtils.class)) {
+      netUtils.when(NetUtils::getHostAddress).thenReturn("10.0.0.1");
+
+      assertEquals(BaseServerStarter.getServerHostname(serverConf), "10.0.0.1");
+
+      netUtils.verify(NetUtils::getHostAddress);
+      netUtils.verify(NetUtils::getHostnameOrAddress, never());
+    }
+  }
+
+  @Test
+  public void testServerHostFallsBackToHostnameWhenConfigured()
+      throws Exception {
+    PinotConfiguration serverConf = new PinotConfiguration();
+    serverConf.setProperty(Helix.SET_INSTANCE_ID_TO_HOSTNAME_KEY, true);
+
+    try (MockedStatic<NetUtils> netUtils = mockStatic(NetUtils.class)) {
+      netUtils.when(NetUtils::getHostnameOrAddress).thenReturn("pinot-server-node1");
+
+      assertEquals(BaseServerStarter.getServerHostname(serverConf), "pinot-server-node1");
+
+      netUtils.verify(NetUtils::getHostnameOrAddress);
+      netUtils.verify(NetUtils::getHostAddress, never());
+    }
+  }
+}

